src/providers/ldap/sdap.c

```diff
@@ -409,92 +414,113 @@ int sdap_parse_entry(TALLOC_CTX *memctx,
 {
     struct sysdb_attrs *attrs;
     BerElement *ber = NULL;
-    struct berval **vals;
+    struct berval bv, *bvals = NULL;
     struct ldb_val v;
-    char *str;
-    int lerrno;
+    int lerrno = 0;
     int i, ret, ai;
     int base_attr_idx = 0;
     const char *name = NULL;
-    bool store;
     bool base64;
+    bool objectClass_matched = false;
     char *base_attr;
     uint32_t range_offset;
     TALLOC_CTX *tmp_ctx = talloc_new(NULL);
-    if (!tmp_ctx) return ENOMEM;
 
-    lerrno = 0;
-    ret = ldap_set_option(sh->ldap, LDAP_OPT_RESULT_CODE, &lerrno);
-    if (ret != LDAP_OPT_SUCCESS) {
-        DEBUG(SSSDBG_MINOR_FAILURE, "ldap_set_option failed [%s], ignored.\n",
-              sss_ldap_err2string(ret));
-    }
+    if (!tmp_ctx) return ENOMEM;
 
     attrs = sysdb_new_attrs(tmp_ctx);
     if (!attrs) {
         ret = ENOMEM;
         goto done;
     }
 
-    str = ldap_get_dn(sh->ldap, sm->msg);
-    if (!str) {
-        ldap_get_option(sh->ldap, LDAP_OPT_RESULT_CODE, &lerrno);
-        DEBUG(SSSDBG_CRIT_FAILURE, "ldap_get_dn failed: %d(%s)\n",
+    lerrno = ldap_get_dn_ber(sh->ldap, sm->msg, &ber, &bv);
+    if (lerrno != LDAP_SUCCESS) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "ldap_get_dn_ber failed: %d(%s)\n",
               lerrno, sss_ldap_err2string(lerrno));
         ret = EIO;
         goto done;
     }
 
-    DEBUG(SSSDBG_TRACE_LIBS, "OriginalDN: [%s].\n", str);
-    PROBE(SDAP_PARSE_ENTRY, "OriginalDN", str, strlen(str));
-    ret = sysdb_attrs_add_string(attrs, SYSDB_ORIG_DN, str);
-    ldap_memfree(str);
-    if (ret) goto done;
+    if (!bv.bv_val || !bv.bv_len) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "ldap_get_dn_ber returned empty DN\n");
+        ret = EIO;
+        goto done;
+    }
+
+    DEBUG(SSSDBG_TRACE_LIBS, "OriginalDN: [%s].\n", bv.bv_val);
+    PROBE(SDAP_PARSE_ENTRY, "OriginalDN", bv.bv_val, bv.bv_len);
+    ret = sysdb_attrs_add_string(attrs, SYSDB_ORIG_DN, bv.bv_val);
+    if (ret)
+        goto done;
+
+    /* If there is no map, skip the objectClass matching logic. */
+    objectClass_matched = !map;
+
+    do {
+        if (bvals) {
+            ber_memfree(bvals);
+            bvals = NULL;
+        }
+        lerrno = ldap_get_attribute_ber(sh->ldap, sm->msg, ber, &bv, &bvals);
+
+        /* Assignments before error-checking, as no de-ref involved... */
+        const char *attr_name = (char *)bv.bv_val;
```

Comment:
No guarantee this octet string is 0-terminated. This cannot be used as a C-string.